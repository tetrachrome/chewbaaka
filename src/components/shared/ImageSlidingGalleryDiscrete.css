/**
 * ImageSlidingGalleryDiscrete.css
 * Chewbaaka
 *
 * Author   : Tomiko
 * Created  : Jul 23, 2020
 * Updated  : Jul 23, 2020
 */

div.ImageSlidingGalleryDiscreteOuterContainer {
  position: relative;
}

div.ImageSlidingGalleryDiscreteOuterContainer_DebugColor {
  background-color: gray;
}

div.ImageSlidingGalleryDiscreteCoreContainer {
}

div.ImageSlidingGalleryDiscreteButtonGroupContainer {
  display: inline-block;
  margin: 20px auto;
}

img.ImageSlidingGalleryDiscreteImgPart {
  filter: drop-shadow(1px 1px 1px #222);
}

div.ImageSlidingGalleryDiscreteButtonContainer {
  position: absolute;
  justify-content: center;
}

div.ImageSlidingGalleryDiscreteButtonLhsContainer {
  left: 0px;
}

div.ImageSlidingGalleryDiscreteButtonRhsContainer {
  right: 0px;
}
