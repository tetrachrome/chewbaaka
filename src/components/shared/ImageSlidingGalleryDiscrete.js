/**
 * ImageSlidingGalleryDiscrete.js
 * Chewbaaka
 *
 * Author   : Tomiko
 * Created  : Jul 23, 2020
 * Updated  : Jul 25, 2020
 */

/**
 * ImageSlidingGalleryDiscrete
 *
 * `ImageSlidingGalleryDiscrete` is a component that illustrates a set of
 * user specified images one at a time, along with a piece of
 * caption string annotating the image.
 *
 * Props:
 *
 *  - `width`: The desired width of the image view.
 *
 *  - `height`: The desired height of the image view. 
 *
 *  - `slides`: An array of objects of { image, caption } tuples.
 */

import React from 'react'

import "semantic-ui-css/semantic.min.css"

import { Button } from "semantic-ui-react"

import './ImageSlidingGalleryDiscrete.css'

import {
  getElementStyleClassName,
  getElementStyleClassNames
} from '../../styling/styling'

import Caption from './Caption'

export default class ImageSlidingGalleryDiscrete extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      activeIndex : 0
    }
    this.handleOnPrevClick = this.handleOnPrevClick.bind(this);
    this.handleOnNextClick = this.handleOnNextClick.bind(this);
    this.handleGridButtonOnClick = this.handleGridButtonOnClick.bind(this);
  }

  handleOnPrevClick() {
    this.setState((prevState) => ({
      activeIndex : Math.max(prevState.activeIndex-1, 0)
    }));
  }

  handleOnNextClick() {
    this.setState((prevState) => ({
      activeIndex : Math.min(prevState.activeIndex+1, this.props.slides.length - 1)
    }));
  }

  handleGridButtonOnClick(idx) {
    this.setState({
      activeIndex : idx
    });
  }

  render() {
    const componentDimensionHeightOffset = 100 + 20;

    const componentDimensionStyle = {
      width: this.props.width,
      height: this.props.height + componentDimensionHeightOffset
    };

    const componentCoreDimensionStyle = {
      width: this.props.width,
      height: this.props.height
    };

    /**
     * Vertically center the prev/next buttons in the center of the image.
     * Need to minus half of the height of the button itself
     * to make it appear really centered.
     */
    const buttonHeight = 43;
    const buttonCommonTopPos = (this.props.height / 2) - (buttonHeight / 2);

    const buttonCommonPositionStyle = {
      top: buttonCommonTopPos
    };

    const buttonItems = this.props.slides.map((_, idx) => (this.renderButtonItem(idx)));

    return (
      <div
        className={getElementStyleClassName("ImageSlidingGalleryDiscreteOuterContainer")}
        style={componentDimensionStyle}
      >
        <div
          className={getElementStyleClassName("ImageSlidingGalleryDiscreteCoreContainer")}
          style={componentDimensionStyle}
        >
          <img className="ImageSlidingGalleryDiscreteImgPart"
            style={componentCoreDimensionStyle}
            src={this.props.slides[this.state.activeIndex].image}
            alt={this.props.slides[this.state.activeIndex].caption}
          />
          <div
            className="ImageSlidingGalleryDiscreteCaptionPart"
            data-testid="ImageSlidingGalleryDiscreteCaptionPart"
          >
            <Caption caption={this.props.slides[this.state.activeIndex].caption} />
          </div>
          <div className={getElementStyleClassName("ImageSlidingGalleryDiscreteButtonGroupContainer")}>
            <Button.Group>
              {buttonItems}
            </Button.Group>
          </div>
        </div>
        <div
          className={getElementStyleClassNames(["ImageSlidingGalleryDiscreteButtonContainer",
                                                "ImageSlidingGalleryDiscreteButtonLhsContainer"])}
          style={buttonCommonPositionStyle}
        >
          <Button
            circular
            icon="arrow alternate circle left outline"
            size="large"
            color="blue"
            disabled={(this.state.activeIndex === 0)}
            onClick={this.handleOnPrevClick}
          />
        </div>
        <div
          className={getElementStyleClassNames(["ImageSlidingGalleryDiscreteButtonContainer",
                                                "ImageSlidingGalleryDiscreteButtonRhsContainer"])}
          style={buttonCommonPositionStyle}
        >
          <Button
            circular
            icon="arrow alternate circle right outline"
            size="large"
            color="blue"
            disabled={(this.state.activeIndex === (this.props.slides.length - 1))}
            onClick={this.handleOnNextClick}
          />
        </div>
      </div>
    );
  }

  renderButtonItem(idx) {
    const color = (this.state.activeIndex === idx) ? "yellow" : "blue";
    return (
      <Button
        key={idx}
        color={color}
        size="huge"
        onClick={(e) => (this.handleGridButtonOnClick(idx))}
      >
        {idx}
      </Button>
    );
  }
}
