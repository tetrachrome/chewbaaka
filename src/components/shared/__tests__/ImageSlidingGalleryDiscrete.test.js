/**
 * ImageSlidingGalleryDiscrete.test.js
 * Chewbaaka
 *
 * Author   : Tomiko
 * Created  : Jul 23, 2020
 * Updated  : Jul 24, 2020
 */

import React from 'react'

import { render, screen } from '@testing-library/react'

import renderer from 'react-test-renderer'

import ImageSlidingGalleryDiscrete from '../ImageSlidingGalleryDiscrete'

const slides=[
  {
    image: '',
    caption: "Researchers used the Tuxtla Statuette to decipher the epi-Olmec writing system, which represents both syllables and words"
  },
  {
    image: '',
    caption: "Listen to a Tsimshian (Pacific Northwest Native) storyteller recount the family history painted on a 38-foot-long house front as specific parts of the design light up."
  }
];

test('renders ImageSlidingGalleryDiscrete component', () => {
  render(
    <ImageSlidingGalleryDiscrete
      slides={slides}
      width={100}
      height={100}
    />
  );

  const captionElement = screen.getByTestId("ImageSlidingGalleryDiscreteCaptionPart");
  expect(captionElement).toBeInTheDocument();
  expect(captionElement.textContent).toBe(slides[0].caption);
});

test('ImageSlidingGalleryDiscrete component snapshot', () => {
  const tree = renderer
    .create(
      <ImageSlidingGalleryDiscrete
        slides={slides}
        width={100}
        height={100}
      />
    ).toJSON();
  expect(tree).toMatchSnapshot();
});
