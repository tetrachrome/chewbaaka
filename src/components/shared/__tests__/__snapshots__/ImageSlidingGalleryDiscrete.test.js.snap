// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSlidingGalleryDiscrete component snapshot 1`] = `
<div
  className="ImageSlidingGalleryDiscreteOuterContainer ImageSlidingGalleryDiscreteOuterContainer_DebugColor DebugWireframe"
  style={
    Object {
      "height": 240,
      "width": 100,
    }
  }
>
  <div
    className="ImageSlidingGalleryDiscreteCoreContainer ImageSlidingGalleryDiscreteCoreContainer_DebugColor DebugWireframe"
    style={
      Object {
        "height": 240,
        "width": 100,
      }
    }
  >
    <img
      alt="Researchers used the Tuxtla Statuette to decipher the epi-Olmec writing system, which represents both syllables and words"
      className="ImageSlidingGalleryDiscreteImgPart"
      src=""
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    />
    <div
      className="ImageSlidingGalleryDiscreteCaptionPart"
      data-testid="ImageSlidingGalleryDiscreteCaptionPart"
    >
      <div
        className="CaptionContainer CaptionContainer_DebugColor DebugWireframe"
        data-testid="CaptionComponentTestId"
        style={
          Object {
            "width": undefined,
          }
        }
      >
        <div
          className="IconContainer IconContainer_DebugColor DebugWireframe"
        >
          <i
            aria-hidden="true"
            className="image icon"
            onClick={[Function]}
          />
        </div>
        <p
          className="ContentPageCaptionTextSize ContentPageCaptionTextSize_DebugColor DebugWireframe CaptionText CaptionText_DebugColor DebugWireframe"
        >
          Researchers used the Tuxtla Statuette to decipher the epi-Olmec writing system, which represents both syllables and words
        </p>
      </div>
    </div>
    <div
      className="ImageSlidingGalleryDiscreteButtonGroupContainer ImageSlidingGalleryDiscreteButtonGroupContainer_DebugColor DebugWireframe"
    >
      <div
        className="ui buttons"
      >
        <button
          className="ui yellow huge button"
          onClick={[Function]}
        >
          0
        </button>
        <button
          className="ui blue huge button"
          onClick={[Function]}
        >
          1
        </button>
      </div>
    </div>
  </div>
  <div
    className="ImageSlidingGalleryDiscreteButtonContainer ImageSlidingGalleryDiscreteButtonContainer_DebugColor DebugWireframe ImageSlidingGalleryDiscreteButtonLhsContainer ImageSlidingGalleryDiscreteButtonLhsContainer_DebugColor DebugWireframe"
    style={
      Object {
        "top": 28.5,
      }
    }
  >
    <button
      className="ui blue large circular icon disabled button"
      disabled={true}
      onClick={[Function]}
      tabIndex={-1}
    >
      <i
        aria-hidden="true"
        className="arrow alternate circle left outline icon"
        onClick={[Function]}
      />
    </button>
  </div>
  <div
    className="ImageSlidingGalleryDiscreteButtonContainer ImageSlidingGalleryDiscreteButtonContainer_DebugColor DebugWireframe ImageSlidingGalleryDiscreteButtonRhsContainer ImageSlidingGalleryDiscreteButtonRhsContainer_DebugColor DebugWireframe"
    style={
      Object {
        "top": 28.5,
      }
    }
  >
    <button
      className="ui blue large circular icon button"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="arrow alternate circle right outline icon"
        onClick={[Function]}
      />
    </button>
  </div>
</div>
`;
